src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/Dependencies/TransitiveDependencyExtensions.cs

```diff
@@ -0,0 +1,149 @@
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        public static Task<bool> IsTransitiveDependencyAsync(this ITransitiveDependencyIdentifier identifier, string packageName, IProject project, CancellationToken token)
+        {
+            if (identifier is null)
+            {
+                throw new ArgumentNullException(nameof(identifier));
+            }
+
+            if (string.IsNullOrEmpty(packageName))
+            {
+                throw new ArgumentException($"'{nameof(packageName)}' cannot be null or empty.", nameof(packageName));
+            }
+
+            if (project is null)
+            {
+                throw new ArgumentNullException(nameof(project));
+            }
+
+            return identifier.IsTransitiveDependencyAsync(packageName, project.GetProjectTranstiveDependencies(), project.TargetFrameworks, token);
+        }
+
+        public static async Task<bool> IsTransitiveDependencyAsync(this ITransitiveDependencyIdentifier identifier, string packageName, IEnumerable<NuGetReference> packages, IEnumerable<TargetFrameworkMoniker> tfms, CancellationToken token)
+        {
+            if (identifier is null)
+            {
+                throw new ArgumentNullException(nameof(identifier));
+            }
+
+            if (string.IsNullOrEmpty(packageName))
+            {
+                throw new ArgumentException($"'{nameof(packageName)}' cannot be null or empty.", nameof(packageName));
+            }
+
+            if (packages is null)
+            {
+                throw new ArgumentNullException(nameof(packages));
+            }
+
+            if (tfms is null)
+            {
+                throw new ArgumentNullException(nameof(tfms));
+            }
+
+            var result = await identifier.GetTransitiveDependenciesAsync(packages, tfms, token).ConfigureAwait(false);
+
+            foreach (var item in result.References)
+            {
+                if (string.Equals(item.Name, packageName, StringComparison.OrdinalIgnoreCase))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private static IEnumerable<NuGetReference> GetProjectTranstiveDependencies(this IProject project)
+        {
+            var set = new HashSet<NuGetReference>();
+            var visited = new HashSet<IProject>();
+
+            var projects = new Queue<IProject>();
+            projects.Enqueue(project);
+
+            do
+            {
+                var current = projects.Dequeue();
+                set.UnionWith(current.PackageReferences);
+
+                foreach (var child in current.ProjectReferences)
+                {
+                    if (visited.Add(child))
+                    {
+                        projects.Enqueue(child);
+                    }
+                }
+            }
+            while (projects.Count > 0);
+
```

Comment:
Is there a perf improvement to this approach compared to our previous approach of `restore` every project and its dependencies?

Reply:
Perf and correctness. 

Perf: Not shelling out to a new process
Correctness: We can do stuff in memory instead of using values from the project file that may be out of date